diff --git a/configure b/configure
index 8f9fb04115..a70c24378d 100755
--- a/configure
+++ b/configure
@@ -1996,8 +1996,6 @@ EXTERNAL_LIBRARY_GPL_LIST="
     libdvdread
     librubberband
     libvidstab
-    libx264
-    libx265
     libxavs
     libxavs2
     libxvid
@@ -2010,6 +2008,8 @@ EXTERNAL_LIBRARY_NONFREE_LIST="
 "
 
 EXTERNAL_LIBRARY_VERSION3_LIST="
+    libx264
+    libx265
     gmp
     libaribb24
     liblensfun
diff --git a/libavutil/riscv/cpu.c b/libavutil/riscv/cpu.c
index 8cc91625c7..481d63a39f 100644
--- a/libavutil/riscv/cpu.c
+++ b/libavutil/riscv/cpu.c
@@ -36,6 +36,7 @@
 #include <sys/hwprobe.h>
 #elif HAVE_ASM_HWPROBE_H
 #include <asm/hwprobe.h>
+#include <asm/unistd.h>
 #include <sys/syscall.h>
 #include <unistd.h>
diff --git a/libavcodec/packet.h b/libavcodec/packet.h
--- a/libavcodec/packet.h
+++ b/libavcodec/packet.h
@@ -630,6 +630,10 @@
      * or muxers.
      */
     AVRational time_base;
+
+    // real time stamp in rtp header
+    int time_real_sec;
+    int time_real_usec;
 } AVPacket;

 #if FF_API_INIT_PACKET
diff --git a/libavformat/rtpdec.c b/libavformat/rtpdec.c
--- a/libavformat/rtpdec.c
+++ b/libavformat/rtpdec.c
@@ -676,6 +676,21 @@
                    s->base_timestamp;
 }

+// user rtp header
+typedef struct _RTP_EX_DATA_
+{
+    unsigned int  magic_no;		// 固定为0x0132DCA9
+    unsigned int  frame_num;	// 帧号
+    unsigned int  frame_type;	// 媒体帧类型：GSF_RTSP_FRAME_TYPE
+    unsigned int  codec;		// 编码类型：GSF_RTSP_CODEC_TYPE
+    unsigned short width;		// 宽度
+    unsigned short height;		// 高度
+    unsigned int   timestamp_sec; // 时间戳，秒
+    unsigned int   timestamp_usec;// 时间戳，微秒
+    unsigned int   size;            	// 媒体帧大小
+}RTP_EX_DATA, * LPRTP_EX_DATA;
+
+
 static int rtp_parse_packet_internal(RTPDemuxContext *s, AVPacket *pkt,
                                      const uint8_t *buf, int len)
 {
@@ -685,6 +700,8 @@
     AVStream *st;
     uint32_t timestamp;
     int rv = 0;
+    int time_real_sec = 0;
+    int time_real_usec = 0;

     csrc         = buf[0] & 0x0f;
     ext          = buf[0] & 0x10;
@@ -736,6 +753,15 @@
         if (len < ext)
             return -1;
         // skip past RTP header extension
+
+        //fill user time info
+        if (len>=32)
+        {
+            RTP_EX_DATA  exData;
+            memcpy(&exData, buf + 4, sizeof(exData));
+            time_real_sec = exData.timestamp_sec;
+            time_real_usec = exData.timestamp_usec;
+        }
         len -= ext;
         buf += ext;
     }
@@ -755,6 +781,8 @@

     // now perform timestamp things....
     finalize_packet(s, pkt, timestamp);
+    pkt->time_real_sec = time_real_sec;
+    pkt->time_real_usec = time_real_usec;

     return rv;
 }
diff --git a/libavformat/demux.c b/libavformat/demux.c
--- a/libavformat/demux.c
+++ b/libavformat/demux.c
@@ -1289,6 +1289,8 @@
         out_pkt->dts          = sti->parser->dts;
         out_pkt->pos          = sti->parser->pos;
         out_pkt->flags       |= pkt->flags & (AV_PKT_FLAG_DISCARD | AV_PKT_FLAG_CORRUPT);
+        out_pkt->time_real_sec = pkt->time_real_sec;
+        out_pkt->time_real_usec = pkt->time_real_usec;

         if (sti->need_parsing == AVSTREAM_PARSE_FULL_RAW)
             out_pkt->pos = sti->parser->frame_offset;