Subject: [PATCH] 将rtp扩展头中的数据解析成时间戳放在AVPacket里
---
Index: libavcodec/packet.h
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/libavcodec/packet.h b/libavcodec/packet.h
--- a/libavcodec/packet.h	(revision d65f4643833429d2128df8bbe8ffb24e9240362b)
+++ b/libavcodec/packet.h	(date 1778056662263)
@@ -630,6 +630,10 @@
      * or muxers.
      */
     AVRational time_base;
+
+    // real time stamp in rtp header
+    int time_real_sec;
+    int time_real_usec;
 } AVPacket;
 
 #if FF_API_INIT_PACKET
Index: libavformat/rtpdec.c
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/libavformat/rtpdec.c b/libavformat/rtpdec.c
--- a/libavformat/rtpdec.c	(revision d65f4643833429d2128df8bbe8ffb24e9240362b)
+++ b/libavformat/rtpdec.c	(date 1778058091175)
@@ -676,6 +676,21 @@
                    s->base_timestamp;
 }
 
+// user rtp header
+typedef struct _RTP_EX_DATA_
+{
+    unsigned int  magic_no;		// 固定为0x0132DCA9
+    unsigned int  frame_num;	// 帧号
+    unsigned int  frame_type;	// 媒体帧类型：GSF_RTSP_FRAME_TYPE
+    unsigned int  codec;		// 编码类型：GSF_RTSP_CODEC_TYPE
+    unsigned short width;		// 宽度
+    unsigned short height;		// 高度
+    unsigned int   timestamp_sec; // 时间戳，秒
+    unsigned int   timestamp_usec;// 时间戳，微秒
+    unsigned int   size;            	// 媒体帧大小
+}RTP_EX_DATA, * LPRTP_EX_DATA;
+
+
 static int rtp_parse_packet_internal(RTPDemuxContext *s, AVPacket *pkt,
                                      const uint8_t *buf, int len)
 {
@@ -685,6 +700,8 @@
     AVStream *st;
     uint32_t timestamp;
     int rv = 0;
+    int time_real_sec = 0;
+    int time_real_usec = 0;
 
     csrc         = buf[0] & 0x0f;
     ext          = buf[0] & 0x10;
@@ -736,6 +753,15 @@
         if (len < ext)
             return -1;
         // skip past RTP header extension
+
+        //fill user time info
+        if (len>=32)
+        {
+            RTP_EX_DATA  exData;
+            memcpy(&exData, buf + 4, sizeof(exData));
+            time_real_sec = exData.timestamp_sec;
+            time_real_usec = exData.timestamp_usec;
+        }
         len -= ext;
         buf += ext;
     }
@@ -755,6 +781,8 @@
 
     // now perform timestamp things....
     finalize_packet(s, pkt, timestamp);
+    pkt->time_real_sec = time_real_sec;
+    pkt->time_real_usec = time_real_usec;
 
     return rv;
 }
Index: libavformat/demux.c
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/libavformat/demux.c b/libavformat/demux.c
--- a/libavformat/demux.c	(revision d65f4643833429d2128df8bbe8ffb24e9240362b)
+++ b/libavformat/demux.c	(date 1778057009472)
@@ -1289,6 +1289,8 @@
         out_pkt->dts          = sti->parser->dts;
         out_pkt->pos          = sti->parser->pos;
         out_pkt->flags       |= pkt->flags & (AV_PKT_FLAG_DISCARD | AV_PKT_FLAG_CORRUPT);
+        out_pkt->time_real_sec = pkt->time_real_sec;
+        out_pkt->time_real_usec = pkt->time_real_usec;
 
         if (sti->need_parsing == AVSTREAM_PARSE_FULL_RAW)
             out_pkt->pos = sti->parser->frame_offset;
